File: Example/UNXNavigatorDemo/Feature/Advanced/Controllers/ViewController08_WebView.m
//
//  ViewController08_WebView.m
//  UNXNavigatorDemo
//
//  Created by Leo Lee on 2020/10/27.
//

#import <WebKit/WebKit.h>
#import <UNXNavigator/UNXNavigator.h>

#import "ViewController08_WebView.h"

#import "UIColor+RandomColor.h"
#import "UIDevice+Additions.h"

@interface ViewController08_WebView () <WKNavigationDelegate, UNXNavigatorInteractable>

@property (nonatomic, strong) WKWebView *webView;
@property (nonatomic, strong) UIProgressView *progressView;
@property (nonatomic, strong) NSURL *requestURL;
@property (nonatomic, strong) UIBarButtonItem *backBarButtonItem;
@property (nonatomic, strong) UIBarButtonItem *closeBarButtonItem;

@end

@implementation ViewController08_WebView

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    if (self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil]) {
        _requestURL = [NSURL URLWithString:@"https://l1dan.gitee.io/"];
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.title = @"Loading...";
    
    if (UIDevice.isPhoneDevice) {
        self.navigationItem.leftBarButtonItems = @[self.backBarButtonItem];
    }
    [self.view addSubview:self.webView];
    [self.unx_navigationBar addSubview:self.progressView];
    [self.webView loadRequest:[NSURLRequest requestWithURL:self.requestURL]];
    
    [self.progressView.topAnchor constraintEqualToAnchor:self.unx_navigationBar.bottomAnchor].active = YES;
    [self.progressView.leftAnchor constraintEqualToAnchor:self.unx_navigationBar.leftAnchor].active = YES;
    [self.progressView.rightAnchor constraintEqualToAnchor:self.unx_navigationBar.rightAnchor].active = YES;
    [self.progressView.heightAnchor constraintEqualToConstant:1.0 / [UIScreen mainScreen].scale].active = YES;
    
    [self.webView.topAnchor constraintEqualToAnchor:self.view.topAnchor].active = YES;
    [self.webView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active = YES;
    [self.webView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor].active = YES;
    [self.webView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = YES;
    
    if (@available(iOS 13.0, *)) {
        self.webView.backgroundColor = [UIColor colorWithDynamicProvider:^UIColor * _Nonnull(UITraitCollection * _Nonnull traitCollection) {
            if (self.view.traitCollection.userInterfaceStyle == UIUserInterfaceStyleDark) {
                self.webView.opaque = NO;
                return [UIColor clearColor];
            }
            self.webView.opaque = YES;
            return [UIColor whiteColor];
        }];
    } else {
        self.webView.opaque = YES;
        self.webView.backgroundColor = nil;
    }
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self.webView addObserver:self forKeyPath:@"estimatedProgress" options:NSKeyValueObservingOptionNew context:NULL];
    [self.webView addObserver:self forKeyPath:@"title" options:NSKeyValueObservingOptionNew context:NULL];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self.webView removeObserver:self forKeyPath:@"estimatedProgress" context:NULL];
    [self.webView removeObserver:self forKeyPath:@"title" context:NULL];
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    [self.progressView setHidden:YES];
}

- (BOOL)unx_useSystemBlurNavigationBar {
    return YES;
}

- (void)dealloc {
    NSLog(@"%s", __FUNCTION__);
}

#pragma mark - Observer

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    if ([keyPath isEqualToString:@"estimatedProgress"] && object == self.webView) {
        [self.progressView setAlpha:1.0f];
        [self.progressView setProgress:self.webView.estimatedProgress animated:YES];
        
        __weak typeof (self)weakSelf = self;
        if(self.webView.estimatedProgress >= 1.0f) {
            [UIView animateWithDuration:0.3 delay:0.3 options:UIViewAnimationOptionCurveEaseOut animations:^{
                [weakSelf.progressView setAlpha:0.0f];
            } completion:^(BOOL finished) {
                [weakSelf.progressView setProgress:0.0f animated:NO];
            }];
        }
    } else if ([keyPath isEqualToString:@"title"] && object == self.webView) {
        self.navigationItem.title = self.webView.title;
    } else {
        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
    }
}

#pragma mark - Action

- (void)clickBackButton:(UIButton *)button {
    if (UIDevice.isPhoneDevice) {
        [self.navigationController unx_triggerSystemBackButtonHandler];
    } else {
        if ([self.webView canGoBack]) {
            [self.webView goBack];
        }
    }
}

- (void)clickCloseButton:(UIButton *)button {
    [self.navigationController popViewControllerAnimated:YES];
}

#pragma mark - Getter

- (UIProgressView *)progressView {
    if (!_progressView) {
        _progressView = [[UIProgressView alloc] init];
        _progressView.trackTintColor = [UIColor customLightGrayColor];
        _progressView.progressTintColor = [UIColor customDarkGrayColor];
        _progressView.translatesAutoresizingMaskIntoConstraints = NO;
    }
    return _progressView;
}

- (WKWebView *)webView {
    if (!_webView) {
        WKWebViewConfiguration *configuration = [[WKWebViewConfiguration alloc] init];
        configuration.preferences.minimumFontSize = 9.0;
        configuration.preferences.javaScriptEnabled = YES;
        
        _webView = [[WKWebView alloc] initWithFrame:CGRectZero configuration:configuration];
        _webView.translatesAutoresizingMaskIntoConstraints = NO;
        _webView.scrollView.decelerationRate = UIScrollViewDecelerationRateNormal;
        _webView.scrollView.keyboardDismissMode = UIScrollViewKeyboardDismissModeOnDrag;
        _webView.allowsBackForwardNavigationGestures = YES;
        _webView.gestureRecognizers.lastObject.enabled = NO;
        _webView.navigationDelegate = self;
    }
    return _webView;
}

- (UIBarButtonItem *)backBarButtonItem {
    if (!_backBarButtonItem) {
        UIButton *customView = [UIButton buttonWithType:UIButtonTypeCustom];
        [customView setImage:[[UIImage imageNamed:@"NavigationBarBack"] imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate] forState:UIControlStateNormal];
        [customView setTintColor:UIColor.customTitleColor];
        [customView sizeToFit];
        [customView addTarget:self action:@selector(clickBackButton:) forControlEvents:UIControlEventTouchUpInside];
        _backBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:customView];
    }
    return _backBarButtonItem;
}

- (UIBarButtonItem *)closeBarButtonItem {
    if (!_closeBarButtonItem) {
        UIButton *customView = [UIButton buttonWithType:UIButtonTypeCustom];
        [customView setImage:[[UIImage imageNamed:@"NavigationBarClose"] imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate] forState:UIControlStateNormal];
        [customView setTintColor:UIColor.customTitleColor];
        [customView sizeToFit];
        [customView addTarget:self action:@selector(clickCloseButton:) forControlEvents:UIControlEventTouchUpInside];
        _closeBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:customView];
    }
    return _closeBarButtonItem;
}

#pragma mark - WKNavigationDelegate

- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation {
    if (UIDevice.isPhoneDevice) {
        if (webView.canGoBack) {
            self.navigationItem.leftBarButtonItems = @[self.backBarButtonItem, self.closeBarButtonItem];
        } else {
            self.navigationItem.leftBarButtonItems = @[self.backBarButtonItem];
        }
    } else {
        if (webView.canGoBack) {
            self.navigationItem.leftBarButtonItems = @[self.backBarButtonItem];
        } else {
            self.navigationItem.leftBarButtonItems = nil;
        }
    }
}

#pragma mark - UNXNavigatorInteractable

- (BOOL)navigationController:(__kindof UINavigationController *)navigationController willPopViewControllerUsingInteractingGesture:(BOOL)interactingGesture {
    if (!interactingGesture && self.webView.canGoBack) {
        [self.webView goBack];
        return NO;
    }
    return YES;
}

@end

File: Example/UNXNavigatorSwiftDemo/Advanced/ViewController08_WebView.swift
//
//  ViewController08_WebView.swift
//  UNXNavigatorSwiftDemo
//
//  Created by Leo Lee on 2020/11/23.
//

import UIKit
import WebKit
import UNXNavigator

class ViewController08_WebView: BaseViewController {
    
    private var titleObservation: NSKeyValueObservation?
    private var progressObservation: NSKeyValueObservation?

    fileprivate lazy var progressView: UIProgressView = {
        let progressView = UIProgressView(frame: .zero)
        progressView.trackTintColor = .customLightGray
        progressView.progressTintColor = .customDarkGray
        progressView.autoresizingMask = [.flexibleWidth, .flexibleTopMargin]
        progressView.translatesAutoresizingMaskIntoConstraints = false
        return progressView
    }()
    
    fileprivate lazy var webView: WKWebView = {
        let configuration = WKWebViewConfiguration()
        configuration.preferences.minimumFontSize = 9.0
        configuration.preferences.javaScriptEnabled = true
        
        let webView = WKWebView(frame: .zero, configuration: configuration)
        webView.scrollView.decelerationRate = .normal
        webView.scrollView.keyboardDismissMode = .onDrag
        webView.allowsBackForwardNavigationGestures = true
        webView.gestureRecognizers?.last?.isEnabled = false
        webView.navigationDelegate = self
        webView.translatesAutoresizingMaskIntoConstraints = false
        return webView
    }()
    
    private lazy var backBarButtonItem: UIBarButtonItem = {
        let customView = UIButton(type: .custom)
        customView.setImage(UIImage(named: "NavigationBarBack")?.withRenderingMode(.alwaysTemplate), for: .normal)
        customView.tintColor = .customTitle
        customView.sizeToFit()
        customView.addTarget(self, action: #selector(clickBackBarButtonItem(_:)), for: .touchUpInside)
        return UIBarButtonItem(customView: customView)
    }()
    
    private lazy var closeBarButtonItem: UIBarButtonItem = {
        let customView = UIButton(type: .custom)
        customView.setImage(UIImage(named: "NavigationBarClose")?.withRenderingMode(.alwaysTemplate), for: .normal)
        customView.tintColor = .customTitle
        customView.sizeToFit()
        customView.addTarget(self, action: #selector(clickCloseBarButtonItem(_:)), for: .touchUpInside)
        return UIBarButtonItem(customView: customView)
    }()
    
    private lazy var requestURL: URL = URL(string: "https://l1dan.gitee.io/")!

    override func viewDidLoad() {
        super.viewDidLoad()

        navigationItem.title = "Loading..."
        
        if UIDevice.isPhone {
            navigationItem.leftBarButtonItems = [self.backBarButtonItem]
        }

        view.addSubview(webView)
        unx_navigationBar.addSubview(progressView)
        webView.load(URLRequest(url: requestURL))

        progressView.topAnchor.constraint(equalTo: unx_navigationBar.bottomAnchor).isActive = true
        progressView.leftAnchor.constraint(equalTo: unx_navigationBar.leftAnchor).isActive = true
        progressView.rightAnchor.constraint(equalTo: unx_navigationBar.rightAnchor).isActive = true
        progressView.heightAnchor.constraint(equalToConstant: 1.0 / UIScreen.main.scale).isActive = true
        
        webView.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
        webView.leftAnchor.constraint(equalTo: view.leftAnchor).isActive = true
        webView.bottomAnchor.constraint(equalTo: view.bottomAnchor).isActive = true
        webView.rightAnchor.constraint(equalTo: view.rightAnchor).isActive = true

        if #available(iOS 13.0, *) {
            webView.backgroundColor = UIColor(dynamicProvider: { [weak self] traitCollection -> UIColor in
                guard let self = self else { return UIColor.white }
                if traitCollection.userInterfaceStyle == .dark {
                    self.webView.isOpaque = false
                    return UIColor.clear
                }
                self.webView.isOpaque = true
                return UIColor.white
            })
        } else {
            webView.isOpaque = true
            webView.backgroundColor = nil
        }
        
        titleObservation = webView.observe(\.title, options: .new) { [weak self] (webView, title) in
            self?.navigationItem.title = webView.title
        }
        
        progressObservation = webView.observe(\.estimatedProgress, options: .new) { [weak self] (webView, estimatedProgress) in
            self?.progressView.alpha = 1.0
            self?.progressView.setProgress(Float(webView.estimatedProgress), animated: true)
            
            if webView.estimatedProgress >= 1.0 {
                UIView.animate(withDuration: 0.3, delay: 0.3, options: .curveEaseOut) {
                    self?.progressView.alpha = 0.0
                } completion: { _ in
                    self?.progressView.setProgress(0.0, animated: false)
                }
            }
        }
    }
    
    override var unx_useSystemBlurNavigationBar: Bool {
        return true
    }
    
    deinit {
        print(#function)
    }
    
}

@objc
extension ViewController08_WebView {
    
    private func clickBackBarButtonItem(_ item: UIBarButtonItem) {
        if UIDevice.isPhone {
            navigationController?.unx_triggerSystemBackButtonHandler()
        } else {
            if webView.canGoBack {
                webView.goBack()
            }
        }
    }
    
    private func clickCloseBarButtonItem(_ item: UIBarButtonItem) {
        navigationController?.popViewController(animated: true)
    }
    
}

extension ViewController08_WebView: WKNavigationDelegate {
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        if UIDevice.isPhone {
            if webView.canGoBack {
                navigationItem.leftBarButtonItems = [backBarButtonItem, closeBarButtonItem]
            } else {
                navigationItem.leftBarButtonItems = [backBarButtonItem]
            }
        } else {
            if webView.canGoBack {
                navigationItem.leftBarButtonItems = [backBarButtonItem]
            } else {
                navigationItem.leftBarButtonItems = nil
            }
        }
    }
}

extension ViewController08_WebView: UNXNavigatorInteractable {
    
    func navigationController(_ navigationController: UINavigationController, willPopViewControllerUsingInteractingGesture interactingGesture: Bool) -> Bool {
        if !interactingGesture && webView.canGoBack {
            webView.goBack()
            return false
        }
        return true
    }
    
}
